bucket.go

```diff
@@ -9,86 +9,113 @@ import (
 	"github.com/libp2p/go-libp2p-core/peer"
 )
 
-// Bucket holds a list of peers.
-type Bucket struct {
+// PeerState is the state of the peer as seen by the Routing Table.
+type PeerState int
+
+const (
+	// PeerStateActive indicates that we know the peer is active/alive.
+	PeerStateActive PeerState = iota
+	// PeerStateMissing indicates that we do not know the state of the peer.
+	PeerStateMissing
+)
+
+// PeerInfo holds all related information for a peer in the K-Bucket.
+type PeerInfo struct {
+	Id    peer.ID
+	State PeerState
+}
+
+// bucket holds a list of peers.
+type bucket struct {
 	lk   sync.RWMutex
 	list *list.List
 }
 
-func newBucket() *Bucket {
-	b := new(Bucket)
+func newBucket() *bucket {
+	b := new(bucket)
 	b.list = list.New()
 	return b
 }
 
-func (b *Bucket) Peers() []peer.ID {
+// returns all peers in the bucket
+// it is safe for the caller to modify the returned objects as it is a defensive copy
+func (b *bucket) peers() []PeerInfo {
+	b.lk.RLock()
+	defer b.lk.RUnlock()
+	var ps []PeerInfo
+	for e := b.list.Front(); e != nil; e = e.Next() {
+		p := e.Value.(*PeerInfo)
+		ps = append(ps, *p)
+	}
+	return ps
+}
+
+// return the Ids of all the peers in the bucket.
+func (b *bucket) peerIds() []peer.ID {
 	b.lk.RLock()
 	defer b.lk.RUnlock()
 	ps := make([]peer.ID, 0, b.list.Len())
 	for e := b.list.Front(); e != nil; e = e.Next() {
-		id := e.Value.(peer.ID)
-		ps = append(ps, id)
+		p := e.Value.(*PeerInfo)
+		ps = append(ps, p.Id)
 	}
 	return ps
 }
 
-func (b *Bucket) Has(id peer.ID) bool {
+// returns the peer with the given Id if it exists
+// returns nil if the peerId does not exist
+func (b *bucket) getPeer(p peer.ID) *PeerInfo {
 	b.lk.RLock()
 	defer b.lk.RUnlock()
```

Comment:
Let's be careful here. I think we need to hold this lock when _mutating_ the peer info, right?

I say "think" because I'm not sure if we even _need_ this lock (anywhere) given that:

* We're not exporting the buckets anymore.
* We appear to be taking a locks in the _actual_ routing table.

However, I haven't checked to make sure it's redundant. If it is redundant, we should remove it. If not, we need to avoid mutating the peer info while not holding the lock.

Reply:
@Stebalien Great point ! The bucket locking is indeed redundant now as all access to the bucket is synchronised via the table lock. 

I've removed the locks and have added some documentation to explain why.